shared/typeinference/codeql/typeinference/internal/TypeInference.qll

```diff
@@ -0,0 +1,783 @@
+               * class Mid<T3> : Base<C<T3>> { }
+               * //    ^^^ `immediateBase`
+               * //        ^^ `tp`
+               * //              ^^^^^^^^^^^ `baseMention`
+               *
+               * class Sub<T4> : Mid<C<T4>> { }
+               * //    ^^^ `sub`
+               * //              ^^^^^^^^^^ `immediateBaseMention`
+               * ```
+               */
+
+              baseTypeMentionHasTypeParameterAt(immediateBase, baseMention, prefix, tp) and
+              t = immediateBaseMention.resolveTypeAt(path0) and
+              path0.startsWith(tp, suffix) and
+              path = prefix.append(suffix)
+            )
+          )
+        )
+      }
+    }
+
+    private import BaseTypes
+
+    /** Provides the input to `Matching`. */
+    signature module MatchingInputSig {
+      /**
+       * A position inside a declaration. For example, the integer position of a
+       * parameter inside a method or the return type of a method.
+       */
+      bindingset[this]
+      class DeclarationPosition {
+        /** Gets a textual representation of this position. */
+        bindingset[this]
+        string toString();
+      }
+
+      /** A declaration, for example a method. */
+      class Declaration {
+        /** Gets a textual representation of this declaration. */
+        string toString();
+
+        /** Gets the location of this declaration. */
+        Location getLocation();
+
+        /** Gets the type parameter at position `tppos` of this declaration, if any. */
+        TypeParameter getTypeParameter(TypeParameterPosition tppos);
+
+        /**
+         * Gets the declared type of this declaration at `path` for position `dpos`.
+         *
+         * For example, if this declaration is the method `int M(bool b)`,
+         * then the declared type at parameter position `0` is `bool` and the
+         * declared return type is `int`.
+         */
+        Type getDeclaredType(DeclarationPosition dpos, TypePath path);
+      }
+
+      /**
+       * A position inside an access. For example, the integer position of an
+       * argument inside a method call.
+       */
+      bindingset[this]
+      class AccessPosition {
+        /** Gets a textual representation of this position. */
+        bindingset[this]
+        string toString();
+      }
+
+      /** An access that targets a declaration, for example a method call. */
+      class Access {
+        /** Gets a textual representation of this access. */
+        string toString();
+
+        /** Gets the location of this access. */
+        Location getLocation();
+
+        /**
+         * Gets the type at `path` for the explicit type argument at position
+         * `tapos` of this access, if any.
+         *
+         * For example, in a method call like `M<int>()`, `int` is an explicit
+         * type argument at position `0`.
+         */
+        Type getExplicitTypeArgument(TypeArgumentPosition tapos, TypePath path);
```

Comment:
I'm a little unsure about the "explicit" here. It makes sense for C# and Rust that have implicit and explicit type arguments. But there are also languages that have implicit type parameters, and we don't have the word "explicit" in `getTypeParameter`. There's also languages that have implicit normal (non-type) arguments, and we don't call normal arguments "explicit". So maybe going for the more generic name without "explicit" is better for a language agnostic interface?